### PR TITLE
try harder to not send duplicate blocks

### PR DESCRIPTION
**Pull request by whyrusleeping**
I havent tested this in the field yet, need to get some better metrics in place first. But this should help out quite a bit with transfer speeds, if my hunch is correct.

### CODE REVIEW

**Comment by jbenet** [comment]
should write a test case for it

**Comment by whyrusleeping** [comment]
@jbenet I've been trying to think about how to work this into a test. The problem i'm having is that we only see the issue on higher latency connections, otherwise the cancels arrive in a timely manner and do their job of preventing duplicate sends.  So finding a way to repro this is what ill be working on today

**Comment by jbenet** [comment]
There's a way to add latency to mocknet or something. Take a look at the epic/ tests.



—

**Comment by jbenet** [comment]
@whyrusleeping  I think the dialer / net stuff should be its own PR?

**Comment by whyrusleeping** [comment]
Yeah, i was just hacking on the same branch. Ill break it up in a little bit

File: exchange/bitswap/bitswap.go [hunk 1]
```diff
@@ -219,6 +219,15 @@ func (bs *Bitswap) HasBlock(ctx context.Context, blk *blocks.Block) error {
 		return errors.New("bitswap is closed")
 	default:
 	}
+	has, err := bs.blockstore.Has(blk.Key())
+	if err != nil {
+		return err
+	}
+
+	if has {
```
Comment (whyrusleeping) [inline exchange/bitswap/bitswap.go]: move this into `ReceiveMessage`, dont log, but maybe record statistics to be printed out by `ipfs bitswap stat`

**Comment by jbenet** [comment]
update here?

**Comment by whyrusleeping** [comment]
@jbenet this is RFCR

**Comment by jbenet** [comment]
@whyrusleeping LGTM

**Pull request merged**